File: android/app/src/main/java/com/douxev/eggshell/ui/correlation/InsightsCard.kt
package com.douxev.eggshell.ui.correlation

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.ArrowDownward
import androidx.compose.material.icons.filled.ArrowUpward
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.res.stringResource
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.douxev.eggshell.R
import com.douxev.eggshell.ui.common.ValueFormat
import com.douxev.eggshell.ui.common.builtinLabelRes
import com.douxev.eggshell.ui.components.CardRule
import com.douxev.eggshell.ui.components.CardVariant
import com.douxev.eggshell.ui.components.EggCard
import com.douxev.eggshell.ui.components.MicroLabel
import com.douxev.eggshell.ui.theme.EggColors
import uniffi.transition.Insight
import uniffi.transition.Valence
import kotlin.math.abs

/**
 * The « Ce qui va ensemble » card: the links the engine found, strongest first.
 *
 * **Every string here is co-occurrence, never cause.** The engine compares
 * averages inside one person's own record over a few weeks; it cannot separate
 * "the missed dose wrecked the sleep" from "the bad week did both". The wording
 * is « va avec » / « les nuits où », never « à cause de » — and the sample
 * counts sit on every row, unhidden, because « 6 nuits contre 9 » and
 * « 40 contre 52 » are very different claims wearing the same sentence.
 *
 * The footer says the same thing in words, once, so the caveat is not carried
 * by the phrasing alone.
 */
@Composable
fun InsightsCard(insights: List<Insight>, modifier: Modifier = Modifier) {
    if (insights.isEmpty()) return

    EggCard(
        variant = CardVariant.Low,
        padding = PaddingValues(horizontal = 18.dp, vertical = 14.dp),
        modifier = modifier,
    ) {
        Text(
            stringResource(R.string.feel_insights_title),
            style = MaterialTheme.typography.titleSmall,
            fontWeight = FontWeight.SemiBold,
            color = MaterialTheme.colorScheme.onSurface,
        )
        MicroLabel(
            stringResource(R.string.feel_insights_sub),
            modifier = Modifier.padding(top = 2.dp),
        )

        insights.take(MAX_SHOWN).forEachIndexed { index, insight ->
            if (index > 0) CardRule(alpha = 0.14f)
            InsightRow(insight)
        }

        CardRule(modifier = Modifier.padding(top = 4.dp), alpha = 0.14f)
        Text(
            stringResource(R.string.feel_insights_caveat),
            style = MaterialTheme.typography.bodySmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.padding(top = 10.dp),
        )
    }
}

@Composable
private fun InsightRow(insight: Insight) {
    val scheme = MaterialTheme.colorScheme
    val rising = insight.delta > 0
    // Neutral metrics get no verdict colour at all: libido and vividness are
    // not achievements, and painting one green would be the app deciding what
    // a good night looks like for somebody else.
    val tint = when {
        insight.valence == Valence.NEUTRAL -> scheme.onSurfaceVariant
        insight.favourable -> EggColors.success
        else -> scheme.error
    }

    Row(
        modifier = Modifier.padding(vertical = 12.dp),
        horizontalArrangement = Arrangement.spacedBy(12.dp),
        verticalAlignment = Alignment.Top,
    ) {
        Icon(
            if (rising) Icons.Filled.ArrowUpward else Icons.Filled.ArrowDownward,
            contentDescription = null,
            modifier = Modifier.size(18.dp),
            tint = tint,
        )
        Column(modifier = Modifier.weight(1f)) {
            Text(
                sentenceFor(insight),
                style = MaterialTheme.typography.bodyMedium,
                color = scheme.onSurface,
            )
            // The counts are the honesty of the row. Never folded away behind a
            // "confidence" badge: the reader can weigh 6-vs-9 themselves, and a
            // single word could not carry the same information.
            MicroLabel(
                stringResource(
                    R.string.feel_insights_sample_fmt,
                    insight.sampleWith,
                    insight.sampleWithout,
                ),
                modifier = Modifier.padding(top = 2.dp),
            )
        }
    }
}

/**
 * « Ton sommeil va avec les jours sans oubli : +2,1 en moyenne. »
 *
 * Built from two halves so the metric and the condition can be translated
 * independently — a sentence assembled from a single format string per
 * combination would be sixty strings, and the sixty-first would be forgotten.
 */
@Composable
private fun sentenceFor(insight: Insight): String {
    val metric = stringResource(metricLabelRes(insight.metricKey))
    val condition = stringResource(conditionLabelRes(insight.againstKey))
    val amount = ValueFormat.significant(abs(insight.delta), digits = 2)
    return stringResource(
        if (insight.delta > 0) R.string.feel_insights_up_fmt else R.string.feel_insights_down_fmt,
        metric,
        condition,
        amount,
    )
}

/**
 * Names the metric exactly as its slider does, by asking the same resolver.
 * A finding that said « Qualité du sommeil » while the editor said
 * `sleep_quality` would read as two unrelated things.
 */
private fun metricLabelRes(key: String): Int {
    val domain = if (key in DREAM_KEYS) "dreams" else "journal"
    val res = builtinLabelRes(domain, key)
    return if (res != 0) res else R.string.feel_insights_metric_generic
}

private val DREAM_KEYS = setOf("sleep_quality", "recall", "vividness", "emotional_tone")

private fun conditionLabelRes(key: String): Int = when (key) {
    "missed_dose" -> R.string.feel_insights_cond_missed
    "late_dose" -> R.string.feel_insights_cond_late
    "clean_day" -> R.string.feel_insights_cond_clean
    "lucid_dream" -> R.string.feel_insights_cond_lucid
    "any_dream" -> R.string.feel_insights_cond_dream
    else -> R.string.feel_insights_metric_generic
}

/**
 * Past this the card stops being a summary. The engine already ranks by size,
 * so the tail is the weakest findings — and a wall of them is how a reader
 * starts believing all of it.
 */
private const val MAX_SHOWN = 5

File: apple/Eggshell/Sources/Correlation/InsightsCard.swift
import SwiftUI
import TransitionCore

/// « Ce qui va ensemble » — the links the engine found, strongest first.
///
/// **Every string here is co-occurrence, never cause.** The engine compares
/// averages inside one person's own record over a few weeks; it cannot separate
/// "the missed dose wrecked the sleep" from "the bad week did both". The
/// wording is « va avec » / « le lendemain », never « à cause de » — and the
/// sample counts sit on every row, unhidden, because « 6 relevés contre 9 » and
/// « 40 contre 52 » are very different claims wearing the same sentence.
///
/// Mirrors Android's `InsightsCard`, including which metrics get no verdict
/// colour: libido and dream intensity are not achievements, and painting one
/// green would be the app deciding what a good night looks like for someone
/// else.
struct InsightsCard: View {
    @Environment(\.palette) private var palette

    let insights: [Insight]

    /// Past this the card stops being a summary. The engine already ranks by
    /// size, so the tail is the weakest findings — and a wall of them is how a
    /// reader starts believing all of it.
    private let maxShown = 5

    var body: some View {
        if insights.isEmpty {
            EmptyView()
        } else {
            EggCard(variant: .low, paddingH: 18, paddingV: 14, spacing: 0) {
                Text("Ce qui va ensemble")
                    .font(EggFont.titleS)
                    .foregroundStyle(palette.onSurface)
                MicroLabel("Repéré dans tes propres relevés, du plus net au moins net.")
                    .padding(.top, 2)

                ForEach(Array(insights.prefix(maxShown).enumerated()), id: \.offset) { pair in
                    if pair.offset > 0 { CardRule(opacity: 0.14) }
                    row(pair.element)
                }

                CardRule(opacity: 0.14).padding(.top, 4)
                Text("Ce sont des coïncidences relevées dans ton journal, pas des causes. Beaucoup d’autres choses bougent en même temps — à regarder comme une piste, pas comme une explication.")
                    .font(EggFont.bodyS)
                    .foregroundStyle(palette.onSurfaceVariant)
                    .fixedSize(horizontal: false, vertical: true)
                    .padding(.top, 10)
            }
        }
    }

    private func row(_ insight: Insight) -> some View {
        let rising = insight.delta > 0
        let tint: Color = {
            // Neutral metrics get no verdict colour at all.
            guard insight.valence != .neutral else { return palette.onSurfaceVariant }
            return insight.favourable ? palette.success : palette.error
        }()

        return HStack(alignment: .top, spacing: Spacing.m) {
            Image(systemName: rising ? "arrow.up" : "arrow.down")
                .font(.system(size: 14, weight: .semibold))
                .foregroundStyle(tint)
                .frame(width: 18)
            VStack(alignment: .leading, spacing: 2) {
                Text(sentence(for: insight))
                    .font(.eggBody)
                    .foregroundStyle(palette.onSurface)
                    .fixedSize(horizontal: false, vertical: true)
                // The counts are the honesty of the row. Never folded away
                // behind a "confidence" badge: the reader can weigh 6-vs-9
                // themselves, and one word could not carry the same thing.
                MicroLabel("Sur \(insight.sampleWith) relevés contre \(insight.sampleWithout)")
            }
            Spacer(minLength: 0)
        }
        .padding(.vertical, 12)
    }

    /// « Ton sommeil est plus haut les jours où tout a été pris à l'heure :
    /// +2,1 en moyenne. »
    ///
    /// Assembled from two halves so the metric and the condition stay
    /// independently translatable — one format string per combination would be
    /// sixty strings, and the sixty-first would be forgotten.
    private func sentence(for insight: Insight) -> String {
        let metric = Self.metricLabel(insight.metricKey)
        let condition = Self.conditionLabel(insight.againstKey)
        let amount = MeasureFormat.value(abs(insight.delta), digits: 2)
        return insight.delta > 0
            ? "\(metric) est plus haut \(condition) : +\(amount) en moyenne."
            : "\(metric) est plus bas \(condition) : −\(amount) en moyenne."
    }

    /// Names the metric exactly as its slider does. A finding that said
    /// « Qualité du sommeil » while the editor said `sleep_quality` would read
    /// as two unrelated things.
    static func metricLabel(_ key: String) -> String {
        switch key {
        case "mood": return "Ton humeur"
        case "dysphoria": return "Ta dysphorie"
        case "euphoria": return "Ton euphorie"
        case "energy": return "Ton énergie"
        case "libido": return "Ta libido"
        case "sleep_quality": return "La qualité de ton sommeil"
        case "recall": return "La netteté de tes souvenirs"
        case "vividness": return "L’intensité de tes rêves"
        case "emotional_tone": return "La tonalité de tes rêves"
        default: return "Cet indicateur"
        }
    }

    static func conditionLabel(_ key: String) -> String {
        switch key {
        case "missed_dose": return "les jours avec une prise oubliée"
        case "late_dose": return "les jours avec une prise en retard"
        case "clean_day": return "les jours où tout a été pris à l’heure"
        case "lucid_dream": return "le lendemain d’un rêve lucide"
        case "any_dream": return "le lendemain d’une nuit avec rêve"
        default: return "dans ces cas"
        }
    }
}
